### DISCUSSION TITLE
Neovim freezes a bit when opening files.

### DISCUSSION THREAD

**Discussion by charliie-dev** [ayamir/nvimdots#300] [Q&A]
I recently notice that my nvim freeze a bit(about 2 sec) when opening any ranbom files, e.g. `nvim plugins.lua`.
However, this kind of "freezing" never shows up when only use `nvim` to open up alpha startup page, it feels **very** snappy, like 0 sec that kind of fast.

For the record, my startup time is ~22ms with my own config, ~40ms with this repo's config. (on 0.9.0-dev)
I'm currently running nvim on Ubuntu22.04(WSL).
I also tested it on 0.9.0-dev and 0.8.0 release, with my own config, but unfortunately both versions has that "2 sec" delay somehow.
I also tried to cut down my plugins form over 100 to ~80, but it still has that kind of delay. (Althogh most of my plugins are lazy-loaded, it shouldn't effect the startup time.)
I tried it on both `Windows Terminal` and `Alacritty`, in and out of `tmux`, all the same.
Profile.log looks pretty normal too. I'm really out of moves.

@ayamir @Jint-lzxy @jczhang02 May I ask for your help? Or do you guys have the same issue like mine?
Here's my [config](https://github.com/CharlesChiuGit/nvimdots.lua). I'll be really appreciate if someone can take a look. Thanks!

**Comment by Jint-lzxy** [comment]
This is because `neovim` is waiting for the corrosponding language server(s) to start _(and respond)_. Try editing a file without any dependent language server, it will be as fast as opening the startup page. I have been trying to adjust lazy-load settings recently _(and more importantly, some settings have been loaded twice, resulting in double startup time!)_. The final goal is to load the corresponding components when necessary.

Hope this can give you some inspiration 👍 

> To test this, you will see a _tremendous_ increase in startup time if you try some servers that may take a long time to start _(for example, `clangd` with a lot of user configuration)_, or limit the network transmission speed.

**Reply by charliie-dev** [reply]
Hmmm. That's a interesting point. But i don't it's the case.
I comment everything in the `completion/plugins.lua`, do a `PackerSync`, remove `_complie.lua` and restart the wsl. The result is still the same.
But it got me thinking, maybe it's something about `ui`. Let me do some experiments.

**Reply by charliie-dev** [reply]
>  I have been trying to adjust lazy-load settings recently (and more importantly, some settings have been loaded twice, resulting in double startup time!).

Really!? Could u give me some example?

**Reply by charliie-dev** [reply]
I think maybe the reason is the `pack.load_compile()` function? After I comment that line, nvim become snappy, tho all the plugins are not loaded lol.

**Reply by ayamir** [reply]
I can't test your config because it will freeze just after `PackerSync` for the first time:(
![image](https://user-images.githubusercontent.com/61657399/197345031-d138f0de-1674-465e-9f52-ed29c8de1980.png)

**Reply by charliie-dev** [reply]
@ayamir Wow! really? Maybe it's still downloading the ts parser?

**Reply by charliie-dev** [reply]
Also the silicon.nvim need rust to build, and it would take some time to build. Maybe that's why?

**Reply by charliie-dev** [reply]
Also i install `node`, `perl` and `ruby` in a custom location, tho i don't think it'll effect packer to download plugins.
![image](https://user-images.githubusercontent.com/32497323/197345569-7e3112af-a295-4e87-8cb2-5ae56e7f2bf2.png)

**Reply by Jint-lzxy** [reply]
> > I have been trying to adjust lazy-load settings recently (and more importantly, some settings have been loaded twice, resulting in double startup time!).
> 
> Really!? Could u give me some example?

I'm testing on my machine and analyzing the files generated by Packer. I'll let you know as soon as I make progress 👍 

> I think maybe the reason is the `pack.load_compile()` function? After I comment that line, nvim become snappy, tho all the plugins are not loaded lol.

Hmm that's interesting... What this function does is to check and load the file compiled by Packer, and generate several necessary user/auto commands. I will look into `_compiled.lua` more carefully. A possible explanation is that if you edit the file directly _(pass it as a command line argument)_, more events will be triggered, causing more plugins to be loaded, thus slowing down startup time 🤔 But it should not have such a significant impact...

**Reply by charliie-dev** [reply]
@ayamir I just remove everything about nvim(cache and plugins) and reinstall everything. Things looks normal on my end? Everything is downloaded as excepted.

**Reply by charliie-dev** [reply]
@Jint-lzxy Interesting! Looking forward of what u can dig out!

**Comment by charliie-dev** [comment]
I just install it in arch(WSL), it still has the 2sec delay. So I think it's not about distros.

**Comment by charliie-dev** [comment]
I think I found one small improvement, which is to stop using anaconda.
By comment out conda-related things in `bashrc/zshrc`, the staruptime improves 3~4ms. Most importantly, toggleterm open up very fast again!

**Comment by charliie-dev** [comment]
@ayamir I think I know why u got freezed while doing `PackerSync`, it's because I set it to ssh connection. 
I also encounter freezing when I'm setting up new machines, but once I switch to https connection, everything runs smoothly and much faster.
